# Table of Content


1. **[Importing Modules (Pandas)](#pandas)**
<br><br> 
2. **[Pandas DataFrame](#dataframes)**
<br><br>
3. **[Manipulating DataFrame](#dataframes)**
<br><br>
4. **[Reading Data from Different Sources](#reading_data)**



<a id="pandas"> </a>
# 1. Pandas

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> Pandas contain data structures and data manipulation tools designed for data cleaning and analysis.
<br><br>
                       Pandas is designed for working with tabular data.<br><br>

A module/library in python is simply a way to organize the code, and it contains either python classes or just functions. 

### Quick look at functions

In [14]:
# what are functions and arguments
def add(x,y):
    c = x+y
    return c

In [15]:
add(200,174)

374

In [ ]:
# what are functions and arguments
def add(x,y= 5):
    c = x+y
    return c

In [12]:
add(4)

9

In [7]:
# what are functions and arguments
def add2(x,z=5):
    c = x+z
    return c

In [13]:
add2(89)

94

In [2]:
#Explain functions and how import works to students
# import inspect
# print(inspect.getsource(add.ad))

In [25]:
#Explain functions and how import works to students
import pandas as pd
import inspect
print(inspect.getsource(pd.isnull))

def isna(obj):
    """
    Detect missing values for an array-like object.

    This function takes a scalar or array-like object and indicates
    whether values are missing (``NaN`` in numeric arrays, ``None`` or ``NaN``
    in object arrays, ``NaT`` in datetimelike).

    Parameters
    ----------
    obj : scalar or array-like
        Object to check for null or missing values.

    Returns
    -------
    bool or array-like of bool
        For scalar input, returns a scalar boolean.
        For array input, returns an array of boolean indicating whether each
        corresponding element is missing.

    See Also
    --------
    notna : Boolean inverse of pandas.isna.
    Series.isna : Detect missing values in a Series.
    DataFrame.isna : Detect missing values in a DataFrame.
    Index.isna : Detect missing values in an Index.

    Examples
    --------
    Scalar arguments (including strings) result in a scalar boolean.

    >>> pd.isna('dog')
    False

    >>> pd.isna(pd.NA)

**How to install and import pandas?**<br>
1. Install pandas:<br><br>
`!pip install pandas`<br><br>
2. Import pandas:<br><br>
`import pandas as pd`

In [29]:
#Check the list of base packages
#!pip list

In [20]:
# install pandas (one time only)
#!pip install pandas

In [10]:
#import pandas 'library/package/modules'


 `as` is used as an alias in pandas. So from now on we will use `pd.` instead of `pandas.` 
 
<br>
<span style="color:crimson">Always use libraries if they are freely available. It saves time, and those codes are already tested, debugged and optimized.</span>

<a id="dataframes"> </a>
# 2. Pandas DataFrames

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> A DataFrame is a tabular representation of data containing an ordered collection of columns, each of which can be a different type (numeric, string, boolean, and so on). <br><br>                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">To read data from a csv file</span>

In [27]:
# read the example.csv file in a dataframe
df = pd.read_csv('BMI.csv')
df

Age  Weight (in kg)  Height (in m)
0    45              60           1.35
1    12              43           1.21
2    54              78           1.50
3    26              65           1.21
4    68              50           1.32
5    21              43           1.52
6    10              32           1.65
7    57              34           1.61
8    75              23           1.24
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

In [28]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             23 non-null     int64  
 1   Weight (in kg)  23 non-null     int64  
 2   Height (in m)   23 non-null     float64
dtypes: float64(1), int64(2)
memory usage: 680.0 bytes


In [43]:
df.isnull?

Signature: df.isnull() -> 'DataFrame'
Docstring:
DataFrame.isnull is an alias for DataFrame.isna.

Detect missing values.

Return a boolean same-sized object indicating if the values are NA.
NA values, such as None or :attr:`numpy.NaN`, gets mapped to True
values.
Everything else gets mapped to False values. Characters such as empty
strings ``''`` or :attr:`numpy.inf` are not considered NA values
(unless you set ``pandas.options.mode.use_inf_as_na = True``).

Returns
-------
DataFrame
    Mask of bool values for each element in DataFrame that
    indicates whether an element is an NA value.

See Also
--------
DataFrame.isnull : Alias of isna.
DataFrame.notna : Boolean inverse of isna.
DataFrame.dropna : Omit axes labels with missing values.
isna : Top-level isna.

Examples
--------
Show which entries in a DataFrame are NA.

>>> df = pd.DataFrame(dict(age=[5, 6, np.NaN],
...                    born=[pd.NaT, pd.Timestamp('1939-05-27'),
...                          pd.Timestamp('1940-04-2

In [30]:
# check the type
type(df)

pandas.core.frame.DataFrame

On checking the data type, we notice it is read as pandas data frame.

## <span style="color:darkgreen;">To print top & bottom rows of the data</span>

In [31]:
#top 3 rows
df.head(3)

Age  Weight (in kg)  Height (in m)
0   45              60           1.35
1   12              43           1.21
2   54              78           1.50

By default, the `.head()` will display **first** five rows. However, we can set the desired number of rows to be displayed.

In [32]:
#bottom 3 rows
df.tail(3)

Age  Weight (in kg)  Height (in m)
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

By default, the `.tail()` will display **last** five rows. However, we can set the desired number of rows to be displayed.

## <span style="color:darkgreen;">To obtain the dimensions of the data</span>

In [33]:
# get the shape of the dataset
df.shape

(23, 3)

## <span style="color:darkgreen;">To know the data types of a data frame</span>

In [34]:
# get all the data types
df.dtypes

Age                 int64
Weight (in kg)      int64
Height (in m)     float64
dtype: object

We see the data type of each variable.

## <span style="color:darkgreen;">Print more information about the data</span>

In [37]:
# All information together
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             23 non-null     int64  
 1   Weight (in kg)  23 non-null     int64  
 2   Height (in m)   23 non-null     float64
dtypes: float64(1), int64(2)
memory usage: 680.0 bytes


We see this output gives the number of rows present in the data `RangeIndex: 23 entries, 0 to 22` There are 23 rows numbered from 0 to 22. And there are a total of three columns - `Data columns (total 3 columns)`. 

Consider `Age 23 non-null int64` indicates that the column named 'Age' has 23 non-null observations having the data type 'int64'

And finally the memory used to save this dataframe is 680 bytes.

In [40]:
# describe your data
df.describe()

Age  Weight (in kg)  Height (in m)
count  23.000000       23.000000      23.000000
mean   44.521739       58.304348       1.528261
std    20.586557       19.401112       0.227309
min    10.000000       21.000000       1.210000
25%    26.000000       44.000000       1.320000
50%    54.000000       65.000000       1.520000
75%    62.000000       76.000000       1.690000
max    75.000000       89.000000       1.850000

# `.loc` and `.iloc` methods

## <span style="color:darkgreen;">Indexing a dataframe using `.loc`</span>

`DataFrame.loc[]` is label-based method, which means that you have to specify rows and columns based on their row and column labels.

In [ ]:
#syntax
dataframe.loc[row-start:row-end, column-start:column-end]

In [45]:
# get first row with all columns
df.loc[0] #takes all the columns

Age               45.00
Weight (in kg)    60.00
Height (in m)      1.35
Name: 0, dtype: float64

In [46]:
# get first row with all columns
df.loc[0, :] #same with the previous one

Age               45.00
Weight (in kg)    60.00
Height (in m)      1.35
Name: 0, dtype: float64

In [47]:
# get the value of Age in the 1st row
df.loc[0, 'Age'] #returns just the 1st row with the 1st column

45

## <span style="color:darkgreen;">Selecting multiple rows</span>

In [54]:
# select row 4, 7, 10
df.loc[[4,7,10]]

Age  Weight (in kg)  Height (in m)
4    68              50           1.32
7    57              34           1.61
10   23              53           1.50

In [55]:
df.loc[3:9]

Age  Weight (in kg)  Height (in m)
3   26              65           1.21
4   68              50           1.32
5   21              43           1.52
6   10              32           1.65
7   57              34           1.61
8   75              23           1.24
9   32              21           1.52

We use two square brackets since we are passing a list of row numbers to be accessed.

## <span style="color:darkgreen;">Selecting a range of rows</span>

In [56]:
# select rows from 11 to 18
df.loc[11:18]

Age  Weight (in kg)  Height (in m)
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85

## <span style="color:darkgreen;">Selecting the first column</span>

In [59]:
# select 1st column with all rows
df.loc[:, 'Age']

0     45
1     12
2     54
3     26
4     68
5     21
6     10
7     57
8     75
9     32
10    23
11    34
12    55
13    23
14    56
15    67
16    26
17    56
18    67
19    26
20    68
21    56
22    67
Name: Age, dtype: int64

To select the last column we use -1, to select the second last column we use -2

## <span style="color:darkgreen;">Select the first two columns</span>

In [60]:
# select the first two columns with all rows
df.loc[:, : 'Weight (in kg)']

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

In [63]:
# select the first two columns with all rows
df.loc[:, :]

Age  Weight (in kg)  Height (in m)
0    45              60           1.35
1    12              43           1.21
2    54              78           1.50
3    26              65           1.21
4    68              50           1.32
5    21              43           1.52
6    10              32           1.65
7    57              34           1.61
8    75              23           1.24
9    32              21           1.52
10   23              53           1.50
11   34              65           1.76
12   55              89           1.65
13   23              45           1.75
14   56              76           1.69
15   67              78           1.85
16   26              65           1.21
17   56              74           1.69
18   67              78           1.85
19   26              65           1.21
20   68              50           1.32
21   56              76           1.69
22   67              78           1.85

## <span style="color:darkgreen;">Indexing a dataframe using `.iloc`</span>

`DataFrame.iloc[]` is integer position-based, so you have to specify rows and columns by their integer position values (0-based integer position).

**Note:** the row names are numbers 

In [64]:
# using iloc select 'Age' & 'Weight (in kg)'
df.iloc[:, 0:2]

Age  Weight (in kg)
0    45              60
1    12              43
2    54              78
3    26              65
4    68              50
5    21              43
6    10              32
7    57              34
8    75              23
9    32              21
10   23              53
11   34              65
12   55              89
13   23              45
14   56              76
15   67              78
16   26              65
17   56              74
18   67              78
19   26              65
20   68              50
21   56              76
22   67              78

<a id="manipulatingDF"> </a>
# 3. Manipulating a Dataframe

<table align="left">
    <tr>
        <td>
            <div align="left", style="font-size:120%">
                <font color="#21618C">
                    <b> CAUTION:<br>
                        1. DataFrame[column] works for any column name, but DataFrame.column only works when the column name is a valid Python variable name.<br>
                        2. New columns cannot be created with the ` data.BMI ` syntax.
                    </b>
                </font>
            </div>
        </td>
    </tr>
</table>

## <span style="color:darkgreen;">Adding a new column to the dataframe</span>

columns/variables/features mean the same

In [65]:
# create a new column BMI which is given by weight / H**2
df['BMI'] = df['Weight (in kg)'] / df['Height (in m)']**2
df

Age  Weight (in kg)  Height (in m)        BMI
0    45              60           1.35  32.921811
1    12              43           1.21  29.369579
2    54              78           1.50  34.666667
3    26              65           1.21  44.395875
4    68              50           1.32  28.696051
5    21              43           1.52  18.611496
6    10              32           1.65  11.753903
7    57              34           1.61  13.116778
8    75              23           1.24  14.958377
9    32              21           1.52   9.089335
10   23              53           1.50  23.555556
11   34              65           1.76  20.983988
12   55              89           1.65  32.690542
13   23              45           1.75  14.693878
14   56              76           1.69  26.609713
15   67              78           1.85  22.790358
16   26              65           1.21  44.395875
17   56              74           1.69  25.909457
18   67              78           1.85  22.790358
19   26              65           1.21  44.395875
20   68              50           1.32  28.696051
21   56              76           1.69  26.609713
22   67              78           1.85  22.790358

In [66]:
# check the shape of the data
df.shape

(23, 4)

In [67]:
y = 69/1.63**2
y

25.970115548195267

## <span style="color:darkgreen;">Adding a new row to the dataframe</span>

In [69]:
# add a new row to the data with values [56, 76, 1.69, 26.609713] using .loc method
df.loc[23] = [56, 76, 1.69, 26.609713]

In [71]:
df.tail()

Age  Weight (in kg)  Height (in m)        BMI
19  26.0            65.0           1.21  44.395875
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
23  56.0            76.0           1.69  26.609713

In [72]:
# add a new row to the data with values [56, 76, 1.69, 26.609713] using .loc method
df.loc[24] = [24, 69, 1.63, 25.970116]
df.tail(3)

Age  Weight (in kg)  Height (in m)        BMI
22  67.0            78.0           1.85  22.790358
23  56.0            76.0           1.69  26.609713
24  24.0            69.0           1.63  25.970116

We see that a new row number 23 has be added to the data.

## <span style="color:darkgreen;">Sorting the dataframe</span>

In [73]:
pd??

Type:        module
String form: <module 'pandas' from 'C:\\Users\\sarah\\anaconda3\\lib\\site-packages\\pandas\\__init__.py'>
File:        c:\users\sarah\anaconda3\lib\site-packages\pandas\__init__.py
Source:     
# flake8: noqa

__docformat__ = "restructuredtext"

# Let users know if they're missing any of our hard dependencies
hard_dependencies = ("numpy", "pytz", "dateutil")
missing_dependencies = []

for dependency in hard_dependencies:
    try:
        __import__(dependency)
    except ImportError as e:
        missing_dependencies.append(f"{dependency}: {e}")

if missing_dependencies:
    raise ImportError(
        "Unable to import required dependencies:\n" + "\n".join(missing_dependencies)
    )
del hard_dependencies, dependency, missing_dependencies

# numpy compat
from pandas.compat import is_numpy_dev as _is_numpy_dev

try:
    from pandas._libs import hashtable as _hashtable, lib as _lib, tslib as _tslib
except ImportError as err:  # pragma: no cover
    module = err.name


In [74]:
# sort the data frame on basis of 'Age' values, by default the values will get sorted in ascending order
# Note: 'ascending = False' will sort the data frame in descending order.
df.sort_values('Age')

Age  Weight (in kg)  Height (in m)        BMI
6   10.0            32.0           1.65  11.753903
1   12.0            43.0           1.21  29.369579
5   21.0            43.0           1.52  18.611496
13  23.0            45.0           1.75  14.693878
10  23.0            53.0           1.50  23.555556
24  24.0            69.0           1.63  25.970116
19  26.0            65.0           1.21  44.395875
3   26.0            65.0           1.21  44.395875
16  26.0            65.0           1.21  44.395875
9   32.0            21.0           1.52   9.089335
11  34.0            65.0           1.76  20.983988
0   45.0            60.0           1.35  32.921811
2   54.0            78.0           1.50  34.666667
12  55.0            89.0           1.65  32.690542
23  56.0            76.0           1.69  26.609713
14  56.0            76.0           1.69  26.609713
17  56.0            74.0           1.69  25.909457
21  56.0            76.0           1.69  26.609713
7   57.0            34.0           1.61  13.116778
15  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
22  67.0            78.0           1.85  22.790358
4   68.0            50.0           1.32  28.696051
20  68.0            50.0           1.32  28.696051
8   75.0            23.0           1.24  14.958377

In [75]:
df.sort_values('Age', ascending = False )

Age  Weight (in kg)  Height (in m)        BMI
8   75.0            23.0           1.24  14.958377
4   68.0            50.0           1.32  28.696051
20  68.0            50.0           1.32  28.696051
22  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
15  67.0            78.0           1.85  22.790358
7   57.0            34.0           1.61  13.116778
14  56.0            76.0           1.69  26.609713
23  56.0            76.0           1.69  26.609713
21  56.0            76.0           1.69  26.609713
17  56.0            74.0           1.69  25.909457
12  55.0            89.0           1.65  32.690542
2   54.0            78.0           1.50  34.666667
0   45.0            60.0           1.35  32.921811
11  34.0            65.0           1.76  20.983988
9   32.0            21.0           1.52   9.089335
16  26.0            65.0           1.21  44.395875
19  26.0            65.0           1.21  44.395875
3   26.0            65.0           1.21  44.395875
24  24.0            69.0           1.63  25.970116
13  23.0            45.0           1.75  14.693878
10  23.0            53.0           1.50  23.555556
5   21.0            43.0           1.52  18.611496
1   12.0            43.0           1.21  29.369579
6   10.0            32.0           1.65  11.753903

In [76]:
df

Age  Weight (in kg)  Height (in m)        BMI
0   45.0            60.0           1.35  32.921811
1   12.0            43.0           1.21  29.369579
2   54.0            78.0           1.50  34.666667
3   26.0            65.0           1.21  44.395875
4   68.0            50.0           1.32  28.696051
5   21.0            43.0           1.52  18.611496
6   10.0            32.0           1.65  11.753903
7   57.0            34.0           1.61  13.116778
8   75.0            23.0           1.24  14.958377
9   32.0            21.0           1.52   9.089335
10  23.0            53.0           1.50  23.555556
11  34.0            65.0           1.76  20.983988
12  55.0            89.0           1.65  32.690542
13  23.0            45.0           1.75  14.693878
14  56.0            76.0           1.69  26.609713
15  67.0            78.0           1.85  22.790358
16  26.0            65.0           1.21  44.395875
17  56.0            74.0           1.69  25.909457
18  67.0            78.0           1.85  22.790358
19  26.0            65.0           1.21  44.395875
20  68.0            50.0           1.32  28.696051
21  56.0            76.0           1.69  26.609713
22  67.0            78.0           1.85  22.790358
23  56.0            76.0           1.69  26.609713
24  24.0            69.0           1.63  25.970116

In [77]:
df.sort_values('Age', ascending = False, inplace = True )

In [78]:
df

Age  Weight (in kg)  Height (in m)        BMI
8   75.0            23.0           1.24  14.958377
4   68.0            50.0           1.32  28.696051
20  68.0            50.0           1.32  28.696051
22  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
15  67.0            78.0           1.85  22.790358
7   57.0            34.0           1.61  13.116778
14  56.0            76.0           1.69  26.609713
23  56.0            76.0           1.69  26.609713
21  56.0            76.0           1.69  26.609713
17  56.0            74.0           1.69  25.909457
12  55.0            89.0           1.65  32.690542
2   54.0            78.0           1.50  34.666667
0   45.0            60.0           1.35  32.921811
11  34.0            65.0           1.76  20.983988
9   32.0            21.0           1.52   9.089335
16  26.0            65.0           1.21  44.395875
19  26.0            65.0           1.21  44.395875
3   26.0            65.0           1.21  44.395875
24  24.0            69.0           1.63  25.970116
13  23.0            45.0           1.75  14.693878
10  23.0            53.0           1.50  23.555556
5   21.0            43.0           1.52  18.611496
1   12.0            43.0           1.21  29.369579
6   10.0            32.0           1.65  11.753903

## <span style="color:darkgreen;">Droping Rows and Columns</span>

In [79]:
# create a copy of the dataframe
df2 = df.copy()
df2

Age  Weight (in kg)  Height (in m)        BMI
8   75.0            23.0           1.24  14.958377
4   68.0            50.0           1.32  28.696051
20  68.0            50.0           1.32  28.696051
22  67.0            78.0           1.85  22.790358
18  67.0            78.0           1.85  22.790358
15  67.0            78.0           1.85  22.790358
7   57.0            34.0           1.61  13.116778
14  56.0            76.0           1.69  26.609713
23  56.0            76.0           1.69  26.609713
21  56.0            76.0           1.69  26.609713
17  56.0            74.0           1.69  25.909457
12  55.0            89.0           1.65  32.690542
2   54.0            78.0           1.50  34.666667
0   45.0            60.0           1.35  32.921811
11  34.0            65.0           1.76  20.983988
9   32.0            21.0           1.52   9.089335
16  26.0            65.0           1.21  44.395875
19  26.0            65.0           1.21  44.395875
3   26.0            65.0           1.21  44.395875
24  24.0            69.0           1.63  25.970116
13  23.0            45.0           1.75  14.693878
10  23.0            53.0           1.50  23.555556
5   21.0            43.0           1.52  18.611496
1   12.0            43.0           1.21  29.369579
6   10.0            32.0           1.65  11.753903

In [80]:
# drop the BMI column
df2.drop('BMI', axis =1, inplace = True)

In [81]:
df2

Age  Weight (in kg)  Height (in m)
8   75.0            23.0           1.24
4   68.0            50.0           1.32
20  68.0            50.0           1.32
22  67.0            78.0           1.85
18  67.0            78.0           1.85
15  67.0            78.0           1.85
7   57.0            34.0           1.61
14  56.0            76.0           1.69
23  56.0            76.0           1.69
21  56.0            76.0           1.69
17  56.0            74.0           1.69
12  55.0            89.0           1.65
2   54.0            78.0           1.50
0   45.0            60.0           1.35
11  34.0            65.0           1.76
9   32.0            21.0           1.52
16  26.0            65.0           1.21
19  26.0            65.0           1.21
3   26.0            65.0           1.21
24  24.0            69.0           1.63
13  23.0            45.0           1.75
10  23.0            53.0           1.50
5   21.0            43.0           1.52
1   12.0            43.0           1.21
6   10.0            32.0           1.65

In [82]:
# dropping the 23rd row
df2.drop(23)

Age  Weight (in kg)  Height (in m)
8   75.0            23.0           1.24
4   68.0            50.0           1.32
20  68.0            50.0           1.32
22  67.0            78.0           1.85
18  67.0            78.0           1.85
15  67.0            78.0           1.85
7   57.0            34.0           1.61
14  56.0            76.0           1.69
21  56.0            76.0           1.69
17  56.0            74.0           1.69
12  55.0            89.0           1.65
2   54.0            78.0           1.50
0   45.0            60.0           1.35
11  34.0            65.0           1.76
9   32.0            21.0           1.52
16  26.0            65.0           1.21
19  26.0            65.0           1.21
3   26.0            65.0           1.21
24  24.0            69.0           1.63
13  23.0            45.0           1.75
10  23.0            53.0           1.50
5   21.0            43.0           1.52
1   12.0            43.0           1.21
6   10.0            32.0           1.65

In [84]:
# reset the index
df2.reset_index()

index   Age  Weight (in kg)  Height (in m)
0       8  75.0            23.0           1.24
1       4  68.0            50.0           1.32
2      20  68.0            50.0           1.32
3      22  67.0            78.0           1.85
4      18  67.0            78.0           1.85
5      15  67.0            78.0           1.85
6       7  57.0            34.0           1.61
7      14  56.0            76.0           1.69
8      23  56.0            76.0           1.69
9      21  56.0            76.0           1.69
10     17  56.0            74.0           1.69
11     12  55.0            89.0           1.65
12      2  54.0            78.0           1.50
13      0  45.0            60.0           1.35
14     11  34.0            65.0           1.76
15      9  32.0            21.0           1.52
16     16  26.0            65.0           1.21
17     19  26.0            65.0           1.21
18      3  26.0            65.0           1.21
19     24  24.0            69.0           1.63
20     13  23.0            45.0           1.75
21     10  23.0            53.0           1.50
22      5  21.0            43.0           1.52
23      1  12.0            43.0           1.21
24      6  10.0            32.0           1.65

## <span style="color:darkgreen;">Droping duplicates</span>

In [86]:
# Check if data has duplicates
df2.duplicated()

8     False
4     False
20     True
22    False
18     True
15     True
7     False
14    False
23     True
21     True
17    False
12    False
2     False
0     False
11    False
9     False
16    False
19     True
3      True
24    False
13    False
10    False
5     False
1     False
6     False
dtype: bool

In [87]:
# Check if data has duplicates
df2.duplicated().sum()

7

In [88]:
# drop duplicates from your data
df2.drop_duplicates(inplace = True)

In [89]:
df2.shape

(18, 3)

## <span style="color:darkgreen;">Checking for missing values</span>

Let's import a new dataset.

In [90]:
# Import missingdata.csv 
mdata = pd.read_csv('missingdata.csv')
mdata.head(2)

Age  Weight (in kg)  Height (in m)
0  45.0            60.0           1.35
1  12.0            43.0           1.21

In [147]:
# check for nulls
mdata.info()
#mdata.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Age             22 non-null     float64
 1   Weight (in kg)  21 non-null     float64
 2   Height (in m)   22 non-null     float64
dtypes: float64(3)
memory usage: 680.0 bytes


The function `.isnull` check whether the data is missing. The `sum()` sums the number of 'True' values in the column. The final output gives the number of missing values in each column.

Here, we see there are 2 missing values in the 'weight' column and one missing value in other columns.

## <span style="color:crimson;">Take home exercise</span>

<a id="reading_data"> </a>
### Reading Data from Different Sources

Note that the files names are used as examples only. You can try importing your own files to execute the below examples.

**1. Read a `.xlsx` file**

`pd.read_excel('example.xlsx')`

**2. Read a `.txt` file**

`data = pd.read_csv('example.txt', sep="\t")`

**3. Read a `.zip` file**

`import zipfile
with zipfile.ZipFile('data.zip') as z:
    with z.open('example.csv') as f:
        file = pd.read_csv(f)
        print(file.head())`

**4. Read a `.html` file**

`df = pd.read_html('example.html', header=1, index_col=0)`

**5. Read a `.json` file**

`pd.read_json('example.json')`